### PR TITLE
MongooseCache resubscription

### PR DESCRIPTION
**Pull request by notsapinho**

### CODE REVIEW

File: src/listener/listener.ts [hunk 1]
```diff
@@ -12,174 +12,188 @@ import WebSocket from "../util/WebSocket";
...
+        ],
+        {
+            onlyEvents: true
+        }
+    );
 
-	this.once("close", () => eventStream.destroy());
+    await eventStream.init();
+    eventStream.on("insert", async (document) => {
+        dispatch.call(this, document);
```
Comment (Flam3rboy) [inline src/listener/listener.ts:40]: why are you changing the stream for every event insert. It should only be changed after the user joins/leaves a guild

**Review by Flam3rboy — CHANGES_REQUESTED** [review]

File: src/listener/listener.ts (hunk 1 above)
Comment (notsapinho) [reply src/listener/listener.ts:40]: oh yeah forgot about that

Comment (notsapinho) [reply src/listener/listener.ts]: fixed

File: src/listener/listener.ts [hunk 2]
```diff
@@ -12,174 +12,188 @@ import WebSocket from "../util/WebSocket";
...
+
+    if(document.event === "GUILD_CREATE") {
+        const newUser = await UserModel.findOne({ id: this.user_id }).lean().exec();
+        var newGuilds = newUser.guilds;
+
+        if (shard_count) {
+            newGuilds = newUser.guilds.filter((x) => (BigInt(x) >> 22n) % shard_count === shard_id);
+        }
+
+        eventStream.changeStream([{ $match: { $or: [{ "fullDocument.guild_id": { $in: newGuilds } }, { "fullDocument.user_id": newUser.id }] } }]);
```
Comment (Flam3rboy) [inline src/listener/listener.ts:62]: can maybe make a function so we don't have to update both pipelines if we update something

File: src/listener/listener.ts [hunk 3]
```diff
@@ -12,174 +12,188 @@ import WebSocket from "../util/WebSocket";
...
+        const channel_id = document.channel_id || document.data?.channel_id;
+        permission = await getPermission(this.user_id, document.guild_id, channel_id);
+    }
+
+    if(document.event === "GUILD_CREATE") {
+        const newUser = await UserModel.findOne({ id: this.user_id }).lean().exec();
+        var newGuilds = newUser.guilds;
+
+        if (shard_count) {
+            newGuilds = newUser.guilds.filter((x) => (BigInt(x) >> 22n) % shard_count === shard_id);
```
Comment (Flam3rboy) [inline src/listener/listener.ts:59]: same

File: src/listener/listener.ts [hunk 4]
```diff
@@ -12,174 +12,188 @@ import WebSocket from "../util/WebSocket";
...
+    console.log("event", document);
+
+    if (document.guild_id) {
+        if (!this.intents.has("GUILDS")) return;
+        const channel_id = document.channel_id || document.data?.channel_id;
+        permission = await getPermission(this.user_id, document.guild_id, channel_id);
+    }
+
+    if(document.event === "GUILD_CREATE") {
+        const newUser = await UserModel.findOne({ id: this.user_id }).lean().exec();
```
Comment (Flam3rboy) [inline src/listener/listener.ts:55]: here you can just use the existing guilds array and push the guild_id in the document

**Review by Flam3rboy — CHANGES_REQUESTED** [review]

File: src/listener/listener.ts (hunk 2 above)
Comment (notsapinho) [reply src/listener/listener.ts:62]: but the changeStream function already do this

Comment (Flam3rboy) [reply src/listener/listener.ts]: I mean a function to get the pipeline

Comment (notsapinho) [reply src/listener/listener.ts]: oh ok

Comment (notsapinho) [reply src/listener/listener.ts]: we could use eventStream.pipeline

**Pull request merged**